scheduling/allocator.rb

```diff
@@ -0,0 +1,253 @@
+          .where(enabled: true)
+          .having { sum(available_storage_gib) >= request.storage_gib }
+          .having { count.function.* >= (request.distinct_storage_devices ? request.storage_volumes.count : 1) })
+
+      ds = ds.where { used_ipv4 < total_ipv4 } if request.ip4_enabled
+      ds = ds.where(vm_host_id: request.host_filter) unless request.host_filter.empty?
+      ds = ds.where(location: request.location_filter) unless request.location_filter.empty?
+      ds.all
+    end
+
+    def self.update_vm(vm_host, vm)
+      ip4, address = vm_host.ip4_random_vm_network if vm.ip4_enabled
+      fail "no ip4 addresses left" if vm.ip4_enabled && !ip4
+      vm.update(
+        vm_host_id: vm_host.id,
+        ephemeral_net6: vm_host.ip6_random_vm_network.to_s,
+        local_vetho_ip: vm_host.veth_pair_random_ip4_addr.to_s,
+        allocated_at: Time.now
+      )
+      AssignedVmAddress.create_with_id(dst_vm_id: vm.id, ip: ip4.to_s, address_id: address.id) if ip4
+      vm.sshable&.update(host: vm.ephemeral_net4 || vm.ephemeral_net6.nth(2))
+    end
+
+    def initialize(host, request)
+      @host = host
+      @request = request
+      @allocations = [
+        HostAllocation.new(:used_cores, host[:total_cores], host[:used_cores], request.cores),
+        HostAllocation.new(:used_hugepages_1g, host[:total_hugepages_1g], host[:used_hugepages_1g], request.mem_gib),
+        StorageAllocation.new(host, request)
+      ]
+      @score = calculate_score
+    end
+
+    def is_valid
+      @allocations.all? { _1.is_valid }
+    end
+
+    def update(vm)
+      vm_host = VmHost[@host[:vm_host_id]]
+      DB.transaction do
+        Allocation.update_vm(vm_host, vm)
+        VmHost.dataset.where(id: @host[:vm_host_id]).update(@allocations.map { _1.update(vm, vm_host) }.reduce(&:merge))
+      end
+    end
+
+    def to_s
+      "#{UBID.from_uuidish(@request.vm_id)} (cpu=#{@request.cores}, mem=#{@request.mem_gib}, storage=#{@request.storage_gib}) -> #{UBID.from_uuidish(@host[:vm_host_id])} (cpu=#{@host[:used_cores]}/#{@host[:total_cores]}, mem=#{@host[:used_hugepages_1g]}/#{@host[:total_hugepages_1g]}, storage=#{@host[:total_storage_gib] - @host[:available_storage_gib]}/#{@host[:total_storage_gib]}), score=#{@score}"
+    end
+
+    private
+
+    def calculate_score
+      util = @allocations.map { _1.utilization }
+
+      # utilization score
+      avg_util = util.sum.fdiv(util.size)
+      utilization_score = @request.target_host_utilization - avg_util
+      utilization_score = utilization_score.abs + 1 if utilization_score < 0
+
+      # imbalance score
+      imbalance_score = util.max - util.min
+
+      # location preference
+      location_preference_score = (@request.location_preference.empty? || @request.location_preference.include?(@host[:location])) ? 0 : 10
```

Comment:
A comment with ranges of each of other scores might be helpful. Then, it will be clearer that 10 is chosen in a way that we give highest priority to location preference.

Reply:
added comments